# Exercise 04 - Due Friday, October 21 at 12pm

*Objectives*: Gain experience loading a CSV dataset into a database model you define yourself and using SQL to explore its contents. Explore the data by writing and executing a number of SQL queries using common syntax and functions and describing your findings.

*Grading criteria*: The tasks should all be completed, and questions should all be answered with SQL queries in the space provided, unless a text answer is requested. Results should be correct, according to the embedded tests. The notebook itself should be completely reproducible; from start to finish, another person should be able to use the same code to obtain the same results as yours.  Note that you will receive no more than partial credit if you do not add text/markdown cells explaining your thinking at each major step in each problem.

*Deadline*: Friday, October 21, 12pm. Submit your notebook to Blackboard and push it to your GitHub repository.

## Setup - obtain data and create database

The US City Open Data Census has data on [restaurant inspections from all over the country](http://us-city.census.okfn.org/dataset/food-safety).  Let's take a look at data from Albuquerque.  *Note*: metadata and other details for this dataset are available at http://data.cabq.gov/business/foodinspections/.

Fyi, to save you a step, I've converted this file into a clean CSV file by using the `dos2unix` command to change its line endings and to switch it to use commas instead of tabs using `csvformat -t inputfile.csv > outputfile.csv`.

First we download the dataset:

In [1]:
!wget https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/exercises/abq-food-inspections-fy2010.csv

--2016-10-24 04:13:25--  https://raw.githubusercontent.com/gwsb-istm-6212-fall-2016/syllabus-and-schedule/master/exercises/abq-food-inspections-fy2010.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.32.133
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.32.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 5025226 (4.8M) [text/plain]
Saving to: ‘abq-food-inspections-fy2010.csv’

abq-food-inspection 100%[=====================>]   4.79M  --.-KB/s   in 0.1s   

2016-10-24 04:13:25 (35.5 MB/s) - ‘abq-food-inspections-fy2010.csv’ saved [5025226/5025226]



## Problem 1 - examine the data (35 points)

Use `csvstat` and other tools as you see fit and observe its contents.  Review the metadata definitions linked to above.  Describe what you see:  What do all the columns mean?  Are there null values?  Which columns interest you the most?  Which columns present some complications or questions you would like to ask?

rename the data with something shorter than the original name

In [2]:
!mv abq-food-inspections-fy2010.csv abq.csv

Here are all variables we have.

In [3]:
!csvcut -n abq.csv

  1: FACILITY_NAME
  2: FACILITY_KEY
  3: SITE_ADDRESS
  4: CITY
  5: STATE
  6: ZIP
  7: OWNER_KEY
  8: OWNER_NAME
  9: NATURE_OF_BUSINESS
 10: STREET_NUMBER
 11: STREET_NAME
 12: STREET_TYPE
 13: POST_DIRECTIONAL
 14: PHONE
 15: PROGRAM_CATEGORY
 16: PROGRAM_CATEGORY_DESCRIPTION
 17: INSPECTION_DATE
 18: INSPECTION_TYPE
 19: INSPECTION_DESC
 20: SERIAL_NUM
 21: ACTION_CODE
 22: ACTION_DESC
 23: RESULT_CODE
 24: RESULT_DESC
 25: VIOLATION_CODE
 26: VIOLATION_DESC
 27: INSPECTION_MEMO


We can take a closer look at the stats of those variables.

In [4]:
!csvstat abq.csv

  1. FACILITY_NAME
	<class 'str'>
	Nulls: False
	Unique values: 2615
	5 most frequent values:
		SMITHS FOOD AND DRUG CENTERS INC:	154
		ISOTOPES PARK - OVATIONS:	93
		PER EHSII CECELIA GARCIA OOB LA REYNA DEL SUR:	85
		ISLETA AMPHITHEATER:	68
		ECHO INC:	68
	Max length: 77
  2. FACILITY_KEY
	<class 'int'>
	Nulls: False
	Min: 13
	Max: 101482
	Sum: 820405773
	Mean: 46730.79135338346
	Median: 50866.0
	Standard Deviation: 32659.794578837962
	Unique values: 2835
	5 most frequent values:
		46323:	93
		65863:	85
		80949:	68
		32817:	68
		91825:	66
  3. SITE_ADDRESS
	<class 'str'>
	Nulls: True
	Unique values: 2456
	5 most frequent values:
		1601 AVENIDA CESAR CHAVEZ SE:	94
		644 OLD COORS BLVD SW:	85
		8601 CENTRAL AV NE:	77
		2200 SUNPORT BLVD SE:	75
		5210 CENTRAL AV SE:	73
	Max length: 35
  4. CITY
	<class 'str'>
	Nulls: False
	Values: ALBUQUERQUE
  5. STATE
	<class 'str'>
	Nulls: False
	Values: 87, NM
  6. ZIP
	<class 'str'>
	Nulls: False
	Unique values: 48
	5 most frequent values:
		87110

###### What do those columns mean?

FACILITY_NAME: Name of the facility where the inspection occurred

FACILITY_KEY: The unique identifier (key) of the facility where the inspection occurred

SITE_ADDRESS: Street address of the facility where the inspection occurred

CITY: City of the facility where the inspection occurred

STATE: State of the facility where the inspection occurred

ZIP:Zip of the facility where the inspection occurred

OWNER_KEY:The unique identifier (key) of the Owner of the facility where the inspection occurred at the time of the inspection

OWNER_NAME: Name of the Owner of the facility where the inspection occurred at the time of the inspection

NATURE_OF_BUSINESS: The type of business, originally for internal use only, provides information about the business. This is not the information provided on the federal tax form

STREET_NUMBER: Street number of the facility where the inspection occurred

STREET_NAME:Street name of the facility where the inspection occurred

STREET_TYPE: Street type of the facility where the inspection occurred

POST_DIRECTIONAL: Postal direction of the facility where the inspection occurred

PHONE:Phone number of the facility where the inspection occurred

PROGRAM_CATEGORY: Id that indicates the type of facility permit the inspection was performed

PROGRAM_CATEGORY_DESCRIPTION: Descript of the id that indicates the type of facility

INSPECTION_DATE: Date the inspection occurred

INSPECTION_TYPE: Code for type of inspection 

INSPECTION_DESC: Description of the inspection

SERIAL_NUM: Unique ID for the inspection daily activity. This is a sequential number assigned chronologically as the data is entered into the system. 

ACTION_CODE: Code of the action taken. If the ACTION_CODE is equal to 00 – NON-GRADED this indicates that the inspector was unable to actually perform an inspection. The RESULT_CODE documents the situation. 

ACTION_DESC: Description of the action taken

RESULT_CODE: Code of the result of the inspection

RESULT_DESC: Description of the result of the inspection

VIOLATION_CODE: The code for the violation

VIOLATION_DESC: Description of the violation that was found

INSPECTION_MEMO: Notes from the inspection made by the inspector


###### Are there null values?

According to the infomation above, there are Null values in : SITE_ADDRESS, NATURE_OF_BUSINESS, STREET_NUMBER, STREET_NAME, STREET_TYPE, POST_DIRECTIONAL, PHONE, PROGRAM_CATEGORY_DESCRIPTION, INSPECTION_TYPE, VIOLATION_CODE, VIOLATION_DESC, and INSPECTION_MEMO. 

###### Which columns interest you the most? 

Those columns that interest me most are description fields such as facility name, nature of business, inspection description, action description, result description, violation description and inspection memo. 

###### Which columns present some complications or questions you would like to ask?

There are a few questions I'd like to ask:
1. Which facilities had the most inspections?
2. Which facilities had the most violations?
3. What are some most frequent violation types or decriptions?
4. Which business type got the most inspections?

Insert a combination of code and text/markdown cells here to review the data and describe what you find.

## Problem 2 - define a database model and load the data (35 points)

Based on what you found above, create and connect to a new database, define a database table in it, and load this dataset into it.  You may use either of the methods for this step you have seen in earlier class notebooks.  You may choose to eliminate variables/columns if they are not relevant or interesting to you - explain your reasoning if you do.  Either way, you should load a majority of the columns present in the source dataset into the database and all of its rows.

Once your data has loaded successfully, run a `COUNT(*)` query to verify that all the data has loaded correctly.

Insert a combination of code and text/markdown cells here to connect to your database, define your table, load its data, and verify the loaded records.

In [5]:
!echo 'redspot' | sudo -S service postgresql restart

[sudo] password for jovyan: Restarting PostgreSQL 9.5 database server: main.


In [6]:
%load_ext sql

/opt/conda/lib/python3.5/site-packages/IPython/config.py:13: ShimWarning: The `IPython.config` package has been deprecated. You should import from traitlets.config instead.
  "You should import from traitlets.config instead.", ShimWarning)
/opt/conda/lib/python3.5/site-packages/IPython/utils/traitlets.py:5: UserWarning: IPython.utils.traitlets has moved to a top-level traitlets package.
  warn("IPython.utils.traitlets has moved to a top-level traitlets package.")


In [7]:
!createdb -U dbuser ex4

In [8]:
%sql postgresql://dbuser@localhost:5432/ex4

'Connected: dbuser@ex4'

In [9]:
!shuf -n 10000 abq.csv | csvstat

  1. GRANT MIDDLE SCHOOL
	<class 'str'>
	Nulls: False
	Unique values: 2479
	5 most frequent values:
		SMITHS FOOD AND DRUG CENTERS INC:	83
		ISOTOPES PARK - OVATIONS:	51
		PER EHSII CECELIA GARCIA OOB LA REYNA DEL SUR:	45
		CIRCLE K 8940:	45
		ECHO INC:	44
	Max length: 77
  2. 2138
	<class 'str'>
	Nulls: False
	Unique values: 2685
	5 most frequent values:
		46323:	51
		91821:	45
		65863:	45
		80949:	44
		91825:	39
	Max length: 12
  3. 1111 EASTERDAY DR NE
	<class 'str'>
	Nulls: True
	Unique values: 2352
	5 most frequent values:
		8601 CENTRAL AV NE:	52
		1601 AVENIDA CESAR CHAVEZ SE:	52
		2200 SUNPORT BLVD SE:	45
		644 OLD COORS BLVD SW:	45
		300 MENAUL NW:	44
	Max length: 35
  4. ALBUQUERQUE
	<class 'str'>
	Nulls: False
	Values: ALBUQUERQUE, CITY
  5. NM
	<class 'str'>
	Nulls: False
	Values: 87, STATE, NM
  6. 87112
	<class 'str'>
	Nulls: False
	Unique values: 47
	5 most frequent values:
		87108:	1124
		87110:	1099
		87102:	896
		87106:	889
		87109:	889
	Max length: 10
  7. 1838
	<cla

In [10]:
%%sql
DROP TABLE IF EXISTS abq;
CREATE TABLE abq (
    FACILITY_NAME VARCHAR(77),
    FACILITY_KEY INTEGER,
    SITE_ADDRESS VARCHAR(35),
    CITY VARCHAR(11),
    STATE VARCHAR(2),
    ZIP VARCHAR(10),
    OWNER_KEY INTEGER,
    OWNER_NAME VARCHAR(82),
    NATURE_OF_BUSINESS VARCHAR(25),
    STREET_NUMBER VARCHAR(7),
    STREET_NAME VARCHAR(24),
    STREET_TYPE VARCHAR(4),
    POST_DIRECTIONAL VARCHAR(2),
    PHONE VARCHAR(15),
    PROGRAM_CATEGORY VARCHAR(4),
    PROGRAM_CATEGORY_DESCRIPTION VARCHAR(40),
    INSPECTION_DATE TIMESTAMP, 
    INSPECTION_TYPE VARCHAR(40),
    INSPECTION_DESC VARCHAR(32),
    SERIAL_NUM VARCHAR(9),
    ACTION_CODE VARCHAR(2),
    ACTION_DESC VARCHAR(50),
    RESULT_CODE VARCHAR(2),
    RESULT_DESC VARCHAR(37),
    VIOLATION_CODE VARCHAR(10),
    VIOLATION_DESC VARCHAR(50),
    INSPECTION_MEMO VARCHAR(40)
    )


Done.
Done.


[]

In [11]:
!pwd

/home/jovyan/work


In [12]:
%%sql
COPY abq FROM '/home/jovyan/work/abq.csv'
CSV
HEADER
QUOTE '"'
DELIMITER ',';

17556 rows affected.


[]

In [13]:
%%sql
SELECT COUNT(*) FROM abq;

1 rows affected.


[(17556,)]

## Problem 3 - explore your data (30 points)

Now that the data is loaded, ask some questions of it!  Identify key variables of interest and note their ranges along with other useful descriptive statistics.  Choose and define a few lines of inquiry, execute queries for each, and describe what you find in your result sets.  Use any query techniques we've seen in class, including aggregate functions, transformations, subqueries, or others as appropriate. 

If you find interesting patterns, adding plots to your exploration would be useful.

Insert a combination of code and text/markdown cells here to explore your data.

In [14]:
%%sql
SELECT column_name, data_type, character_maximum_length, is_nullable
FROM INFORMATION_SCHEMA.COLUMNS WHERE table_name = 'abq';

27 rows affected.


[('facility_name', 'character varying', 77, 'YES'),
 ('facility_key', 'integer', None, 'YES'),
 ('site_address', 'character varying', 35, 'YES'),
 ('city', 'character varying', 11, 'YES'),
 ('state', 'character varying', 2, 'YES'),
 ('zip', 'character varying', 10, 'YES'),
 ('owner_key', 'integer', None, 'YES'),
 ('owner_name', 'character varying', 82, 'YES'),
 ('nature_of_business', 'character varying', 25, 'YES'),
 ('street_number', 'character varying', 7, 'YES'),
 ('street_name', 'character varying', 24, 'YES'),
 ('street_type', 'character varying', 4, 'YES'),
 ('post_directional', 'character varying', 2, 'YES'),
 ('phone', 'character varying', 15, 'YES'),
 ('program_category', 'character varying', 4, 'YES'),
 ('program_category_description', 'character varying', 40, 'YES'),
 ('inspection_date', 'timestamp without time zone', None, 'YES'),
 ('inspection_type', 'character varying', 40, 'YES'),
 ('inspection_desc', 'character varying', 32, 'YES'),
 ('serial_num', 'character varying', 9, 'YES'),
 ('action_code', 'character varying', 2, 'YES'),
 ('action_desc', 'character varying', 50, 'YES'),
 ('result_code', 'character varying', 2, 'YES'),
 ('result_desc', 'character varying', 37, 'YES'),
 ('violation_code', 'character varying', 10, 'YES'),
 ('violation_desc', 'character varying', 50, 'YES'),
 ('inspection_memo', 'character varying', 40, 'YES')]

First, let's find out what's the most common reason for a inspection. 

In [15]:
%%sql
SELECT inspection_desc, COUNT(*) AS count
FROM abq
GROUP BY inspection_desc
ORDER BY count DESC

15 rows affected.


[('ROUTINE FOOD INSPECTION', 12439),
 ('EMERGENCY RESPONSE', 1987),
 ('New Business Inspection', 1005),
 ('PRE-OPENING FOOD INSPECTION', 743),
 ('ADDITIONAL SERVICE PROVIDED', 439),
 ('MOBILE UNIT FOOD INSPECTION', 377),
 ('PHONE CALL', 247),
 ('FOOD FOLLOW-UP INSPECTION', 143),
 ('FOLLOW-UP DOCUMENTATION RECEIVED', 75),
 ('MANAGERIAL ASSISTANCE', 54),
 ('IMMINENT HEALTH HAZARD', 36),
 ('FIRE CALL', 5),
 ('PLAN REVIEW INSPECTION', 4),
 ('COMPLAINT FOOD INSPECTION', 1),
 ('E-MAIL', 1)]

We can see from the result that there are 15 reasons or descriptions for a inspection and the large portion of the inspections are just routine food inspection. 

Then let's take a look at which facilites got the most inspections. 

In [16]:
%%sql
SELECT facility_name, COUNT(*) AS count
FROM abq
WHERE inspection_desc IN (
    SELECT inspection_desc
    FROM abq
    GROUP BY inspection_desc
    ORDER BY COUNT(*) DESC
    LIMIT 1)
GROUP BY facility_name
ORDER BY count DESC
LIMIT 10;

10 rows affected.


[('SMITHS FOOD AND DRUG CENTERS INC', 130),
 ('ISOTOPES PARK - OVATIONS', 93),
 ('ISLETA AMPHITHEATER', 68),
 ('ECHO INC', 65),
 ('OOB  PROS RANCH MARKETS OF ALBUQUERQUE', 52),
 ('TALIN MARKET WORLD FOOD FARE', 52),
 ('MCDONALDS', 47),
 ('SUBWAY', 46),
 ('OOB SUNFLOWER FARMERS MARKET', 42),
 ('LA PETITE ACADEMY', 41)]

It's surprising to me that some restaurants can get as many as 130 inspections. 

I'm also curious about what are some most common violation descriptions. 

In [17]:
%%sql
SELECT violation_desc, COUNT(*) AS count
FROM abq
WHERE inspection_desc IN (
    SELECT inspection_desc
    FROM abq
    GROUP BY inspection_desc
    ORDER BY COUNT(*) DESC
    LIMIT 1)
GROUP BY violation_desc
ORDER BY count DESC
LIMIT 10;

10 rows affected.


[('Additional Comments', 5052),
 ('No Violations Found', 3048),
 ('Plumbing: Improperly installed/maintained/supplied', 445),
 ('Non-Food contact surfaces', 407),
 ('Improper sanitizing', 398),
 ('Floors, Walls, Ceilings', 373),
 ('Food unprotected', 370),
 ('Food contact surfaces', 230),
 ('Cold Holding', 178),
 ('Thermometers not provided, not accurate/consp', 154)]

There is a large portion of violation descriptions that needs furthur interpretion. 

We can also find out which facilities had a most violations. 

In [18]:
%%sql
SELECT facility_name, COUNT(*) AS count
FROM abq
WHERE violation_desc IN (
    SELECT violation_desc
    FROM abq
    GROUP BY violation_desc
    ORDER BY COUNT(*) DESC
    LIMIT 1)
GROUP BY facility_name
ORDER BY count DESC
LIMIT 10;

10 rows affected.


[('SMITHS FOOD AND DRUG CENTERS INC', 64),
 ('ISOTOPES PARK - OVATIONS', 46),
 ('PER EHSII CECELIA GARCIA OOB LA REYNA DEL SUR', 43),
 ('ISLETA AMPHITHEATER', 34),
 ('TALIN MARKET WORLD FOOD FARE', 27),
 ('OOB  PROS RANCH MARKETS OF ALBUQUERQUE', 25),
 ('SUBWAY', 24),
 ('HOT DOG MAN & CO', 21),
 ('FACILITY OOB 2/28/14 PER EHSII RANDY MARTINEZ THE ICE CREAM MAN', 21),
 ('WHOLE FOODS MARKET', 21)]

It's no surprise that some restaurants that have most inspections also have the most violations. This in return shows that those restaurants do need to be inspect more often. 

I also want to know about what are some most common business type to get the most inspections. 

In [19]:
%%sql
SELECT PROGRAM_CATEGORY_DESCRIPTION, COUNT(*) AS count
FROM abq
WHERE violation_desc IN (
    SELECT violation_desc
    FROM abq
    GROUP BY violation_desc
    ORDER BY COUNT(*) DESC
    LIMIT 1)
GROUP BY PROGRAM_CATEGORY_DESCRIPTION
ORDER BY count DESC
LIMIT 10;

10 rows affected.


[('Food Service Establishment', 2694),
 ('Retail -Grocery', 461),
 ('School', 352),
 ('Bar', 344),
 ('Retail - Prepackage Only', 342),
 ('Childcare', 296),
 ('Retail - Self Service & Prepackage Foods', 252),
 ('Food Processor', 201),
 ('Mobile Pre-Packaged', 181),
 ('Institutional Kitchen', 146)]

Sometimes it can be particularly useful to extract all kinds of dates out into variables, maybe we can find out some important patterns.

In [20]:
%%sql
SELECT TO_CHAR(inspection_date, 'YYYY-MM-DD') AS day_of_year, 
    TO_CHAR(inspection_date, 'YYYY') AS year,
    TO_CHAR(inspection_date, 'MM') AS month,
    TO_CHAR(inspection_date, 'DD') AS day,
    TO_CHAR(inspection_date, 'Day') AS day_of_week_str,
    TO_CHAR(inspection_date, 'D') AS day_of_week,
    CASE WHEN CAST(TO_CHAR(inspection_date, 'D') AS INTEGER) >= 6 
        THEN 1 
        ELSE 0
    END AS is_weekend,
    CASE WHEN CAST(TO_CHAR(inspection_date, 'D') AS INTEGER) < 6 
        THEN 1 
        ELSE 0
    END AS is_weekday,
    TO_CHAR(inspection_date, 'HH24') AS hour_24,
    TO_CHAR(inspection_date, 'Q') AS quarter,
    inspection_desc as inspection_desc
FROM abq
LIMIT 10;

10 rows affected.


[('2009-10-21', '2009', '10', '21', 'Wednesday', '4', 0, 1, '00', '4', 'EMERGENCY RESPONSE'),
 ('2009-10-21', '2009', '10', '21', 'Wednesday', '4', 0, 1, '00', '4', 'ROUTINE FOOD INSPECTION'),
 ('2009-10-21', '2009', '10', '21', 'Wednesday', '4', 0, 1, '00', '4', 'ROUTINE FOOD INSPECTION'),
 ('2010-04-14', '2010', '04', '14', 'Wednesday', '4', 0, 1, '00', '2', 'ROUTINE FOOD INSPECTION'),
 ('2010-04-14', '2010', '04', '14', 'Wednesday', '4', 0, 1, '00', '2', 'ROUTINE FOOD INSPECTION'),
 ('2010-03-23', '2010', '03', '23', 'Tuesday  ', '3', 0, 1, '00', '1', 'ROUTINE FOOD INSPECTION'),
 ('2010-03-23', '2010', '03', '23', 'Tuesday  ', '3', 0, 1, '00', '1', 'ROUTINE FOOD INSPECTION'),
 ('2010-03-23', '2010', '03', '23', 'Tuesday  ', '3', 0, 1, '00', '1', 'ROUTINE FOOD INSPECTION'),
 ('2009-12-15', '2009', '12', '15', 'Tuesday  ', '3', 0, 1, '00', '4', 'EMERGENCY RESPONSE'),
 ('2009-12-15', '2009', '12', '15', 'Tuesday  ', '3', 0, 1, '00', '4', 'ROUTINE FOOD INSPECTION')]

In [22]:
%%sql
DROP TABLE IF EXISTS days;
CREATE TABLE days_abq (
    id SERIAL,
    day_of_year CHAR(10),
    year INTEGER,
    month INTEGER,
    day INTEGER,
    day_of_week_str CHAR(9),
    day_of_week INTEGER,
    is_weekend BOOLEAN,
    is_weekday BOOLEAN,
    hour_24 INTEGER,
    quarter INTEGER,
    inspection_desc char(50)
);


Done.
Done.


[]

In [24]:
%%sql
INSERT INTO days_abq (day_of_year, year, month, day, day_of_week_str, day_of_week,
                  is_weekend, is_weekday, hour_24, quarter, inspection_desc)
SELECT DISTINCT TO_CHAR( inspection_date, 'YYYY-MM-DD') AS day_of_year, 
    CAST(TO_CHAR( inspection_date, 'YYYY') AS INTEGER) AS year,
    CAST(TO_CHAR( inspection_date, 'MM') AS INTEGER) AS month,
    CAST(TO_CHAR( inspection_date, 'DD') AS INTEGER) AS day,
    TO_CHAR( inspection_date, 'Day') AS day_of_week_str,
    CAST(TO_CHAR( inspection_date, 'D') AS INTEGER) AS day_of_week,
    CASE WHEN CAST(TO_CHAR( inspection_date, 'D') AS INTEGER) IN (1, 7) 
        THEN TRUE
        ELSE FALSE
    END AS is_weekend,
    CASE WHEN CAST(TO_CHAR( inspection_date, 'D') AS INTEGER) NOT IN (1, 7) 
        THEN TRUE
        ELSE FALSE
    END AS is_weekday,
    CAST(TO_CHAR( inspection_date, 'HH24') AS INTEGER) AS hour_24,
    CAST(TO_CHAR( inspection_date, 'Q') AS INTEGER) AS quarter,
    inspection_desc as inspection
FROM abq;


1204 rows affected.


[]

Now we can take a look as how the inspections occurred in relation to dates and month. 

In [27]:
%%sql
select month, count(*) as month_count
from days_abq
where inspection_desc not in ('No Violations Found', 'None','Additional Comments', 'None', 'Delivered Informational Brochure')
group by month
order by month_count desc
limit 10;

10 rows affected.


[(12, 131),
 (9, 123),
 (10, 119),
 (11, 108),
 (1, 103),
 (3, 101),
 (8, 99),
 (7, 92),
 (6, 91),
 (4, 86)]

## Bonus (10 points)

We have seen a few simple techniques for normalizing data in class.  Choose a column or set of columns that might be redundant and normalize them out to another table:

* Identify the columns to normalize in the source data table
* Create one or more new tables to hold the normalized records
* Extract data into the new tables
* Re-insert the new foreign key references from the new table into the source table

Through previous work, I think there are redundancy in facility_name and owner_name. I'm going to normalize those variables. 

In [28]:
%%sql
SELECT DISTINCT facility_name, facility_key
FROM abq
ORDER BY facility_name
limit 10;

10 rows affected.


[('00BSAY IT WITH CAKE', 50534),
 ('2000 VIETNAM RESTAURANT', 31468),
 ('328 CHINESE CUISINE', 72789),
 ('5 STAR BURGERS', 90929),
 ('66 DINER', 7820),
 ('7 BAR ELEMENTARY SCHOOL', 2207),
 ('7 ELEVEN', 21347),
 ('7 ELEVEN 57722', 21346),
 ('7 ELEVEN 700', 21348),
 ('7 ELEVEN 702', 21350)]

In [29]:
%%sql
SELECT DISTINCT owner_name, owner_key
FROM abq
ORDER BY owner_name
limit 10;

10 rows affected.


[('110 SUNPORT LLC', 80806),
 ('1ST CLASS LEARNING CENTER', 39984),
 ('2000 VIETMAN RESTAURANT', 29294),
 ('21ST CENTURY PUBLIC ACADEMY', 65277),
 ('300 CLUB GRILL', 80954),
 ('328 CHINESE CUISINE', 66644),
 ('3RD ALARM COFFEE', 67195),
 ('5 STAR ABQ INC', 82383),
 ('6001 INC', 15660),
 ('66 DINER', 6743)]

In [32]:
%%sql
SELECT restaurant_name, restaurant_id
FROM (
    SELECT DISTINCT facility_name as restaurant_name, facility_key AS restaurant_id FROM abq
    UNION
    SELECT DISTINCT owner_name as restaurant_name, owner_key AS restaurant_id FROM abq
    ) AS d
LIMIT 10;

10 rows affected.


[('OOB SPECIALTY HOSPITAL OF ALBUQUER', 91957),
 ('QUIZNOS SUB 12316', 82796),
 ('QUALITY INN UNIVERSITY', 92080),
 ('OOB WEENIE WAGON', 85431),
 ('SUSHI GEN', 54063),
 ('FEDERICO SR', 77232),
 ('SMITHS 439', 4291),
 ('SB RESTAURANT CO', 61877),
 ('DUKE CITY AQUATICS FOUNDATION', 79374),
 ('IL VICINO HEIGHTS', 29317)]

In [37]:
%%sql
DROP TABLE IF EXISTS names;
CREATE TABLE names(
    id SERIAL,
    name VARCHAR(180),
    restaurant_id INTEGER
);

Done.
Done.


[]

In [38]:
%%sql
INSERT INTO names (name, restaurant_id)
SELECT restaurant_name, restaurant_id FROM (
   SELECT DISTINCT facility_name as restaurant_name, facility_key AS restaurant_id FROM abq
    UNION
    SELECT DISTINCT owner_name as restaurant_name, owner_key AS restaurant_id FROM abq
) AS d;

4727 rows affected.


[]

In [44]:
%%sql
ALTER TABLE abq 
ADD COLUMN facility_newid INTEGER;

Done.


[]

In [45]:
%%sql
UPDATE abq as r
SET facility_newid = n.id
FROM names as n
WHERE r.facility_name = n.name;

17556 rows affected.


[]

In [46]:
%%sql
ALTER TABLE abq 
ADD COLUMN owner_newid INTEGER;

Done.


[]

In [47]:
%%sql
UPDATE abq as r
SET owner_newid = n.id
FROM names as n
WHERE r.owner_name = n.name;

17556 rows affected.


[]

In [48]:
%%sql
select distinct facility_name,owner_name, facility_newid, owner_newid 
from abq 
limit 10;

10 rows affected.


[('OOB ACE SUSHI', 'ASIANA MANAGEMENT GROUP', 2791, 4710),
 ('OOB CAFE CHORONI', 'NEMESIO MORANTES', 563, 411),
 ('OOB NEW CHINA WOK INC', 'NEW CHINA WOK INC', 2773, 2075),
 ('OOB BUSINESS SOLD QUALITY INN & ECONOLODGE', 'OOB BUSINESS SOLD ASN LLC', 1168, 713),
 ('OOB GRAND CHINA SUPER BUFFET', 'OOB B/W BUSINESS CLOSED GUO AND CHEN INC', 924, 3228),
 ('CANTEEN OF NM @ HOPE CHRISTIAN ES', 'CANTEEN SNACK BAR', 1918, 98),
 ('CARNICERIA LA ESPECIAL', 'SAUL MANRIQUEZ', 1450, 4420),
 ('OOB 03/20/12ROCKY MTN CHOCOLATE FACTORY', 'OOB 03/20/12SABROSO SWEETS LLC PER WRITTEN REQUEST', 1537, 971),
 ('FRONTIER RESTAURANT', 'FRONTIER RESTAURANT', 4653, 4653),
 ('MANZANO DECA STORE', 'MANZANO HIGH SCHOOL DECA', 100, 112)]